src/core/utils/global_address.cpp

```diff
@@ -0,0 +1,264 @@
+
+    VerifyOrExit(aContext != NULL, error = kThreadError_InvalidArgs);
+
+    for (uint32_t i = 0; i <= 255; i++)
+    {
+        error = createIid(aAddress, data);
+        VerifyOrExit(error == kThreadError_Failed,);
+
+        data->mDadCounter++;
+    }
+
+exit:
+    return error;
+}
+
+ThreadError SemanticallyOpaqueIidGenerator::createIid(otNetifAddress *aAddress,
+                                                      otSemanticallyOpaqueIidGeneratorData *aData)
+{
+    ThreadError error = kThreadError_None;
+    Crypto::Sha256 sha256;
+    uint8_t hash[Crypto::Sha256::kHashSize];
+
+    sha256.Start();
+
+    sha256.Update(aAddress->mAddress.mFields.m8, aAddress->mPrefixLength / 8);
+
+    VerifyOrExit(aData->mInterfaceId != NULL, error = kThreadError_InvalidArgs);
+    sha256.Update(aData->mInterfaceId, aData->mInterfaceIdLength);
+
+    if (aData->mNetworkIdLength)
+    {
+        VerifyOrExit(aData->mNetworkId != NULL, error = kThreadError_InvalidArgs);
+        sha256.Update(aData->mNetworkId, aData->mNetworkIdLength);
+    }
+
+    sha256.Update((uint8_t *) &aData->mDadCounter, sizeof(aData->mDadCounter));
+
+    VerifyOrExit(aData->mSecretKey != NULL, error = kThreadError_InvalidArgs);
+    sha256.Update(aData->mSecretKey, aData->mSecretKeyLength);
+
+    sha256.Finish(hash);
+
+    memcpy(&aAddress->mAddress.mFields.m8[8], &hash[sizeof(hash) - 8], 8);
+
+    VerifyOrExit(!IsAddressRegistered(aAddress, aData->mAddresses), error = kThreadError_Failed);
+    VerifyOrExit(!IsAddressReserved(aAddress), error = kThreadError_Failed);
+
+exit:
+    return error;
+}
+
+bool SemanticallyOpaqueIidGenerator::IsAddressRegistered(otNetifAddress *aCreatedAddress, otNetifAddress *aAddressList)
+{
+    otNetifAddress *address = aAddressList;
+
+    while (address != NULL)
+    {
+        if (0 == memcmp(aCreatedAddress->mAddress.mFields.m8, address->mAddress.mFields.m8,
+                        sizeof(address->mAddress.mFields)))
+        {
+            return true;
+        }
+
+        address = address->mNext;
+    }
+
+    return false;
+}
+
+bool SemanticallyOpaqueIidGenerator::IsAddressReserved(otNetifAddress *aCreatedAddress)
+{
+    // Subnet-Router Anycast
+    if ((aCreatedAddress->mAddress.mFields.m32[2] == 0) &&
+        (aCreatedAddress->mAddress.mFields.m32[3] == 0))
+    {
+        return true;
+    }
+
+    // Reserved Subnet Anycast Addresses
+    if ((aCreatedAddress->mAddress.mFields.m8[8]  == 0xfd) &&
+        (aCreatedAddress->mAddress.mFields.m8[9]  == 0xff) &&
+        (aCreatedAddress->mAddress.mFields.m8[10] == 0xff) &&
+        (aCreatedAddress->mAddress.mFields.m8[11] == 0xff) &&
+        (aCreatedAddress->mAddress.mFields.m8[12] == 0xff) &&
+        (aCreatedAddress->mAddress.mFields.m8[13] == 0xff) &&
+        (aCreatedAddress->mAddress.mFields.m8[14] == 0xff) &&
+        (aCreatedAddress->mAddress.mFields.m8[15] >= 0x80))
+    {
+        return true;
```

Comment:
There is a lot of style bugs. Is the coding standard spec available to prevent this kind of problems in the future?